src/main/kotlin/com/sourcegraph/cody/chat/AgentChatSession.kt

```diff
@@ -105,103 +109,122 @@ private constructor(
             text,
             displayText,
         )
-    addMessageAtIndex(humanMessage, index = messages.count(), shouldAddBlinkingCursor = null)
+    addMessageAtIndex(humanMessage, index = messages.count())
 
     val responsePlaceholder =
         ChatMessage(
             Speaker.ASSISTANT,
             text = "",
             displayText = "",
         )
-    addMessageAtIndex(responsePlaceholder, index = messages.count(), shouldAddBlinkingCursor = true)
+    addMessageAtIndex(responsePlaceholder, index = messages.count())
 
     submitMessageToAgent(humanMessage, contextItems)
   }
 
   private fun submitMessageToAgent(humanMessage: ChatMessage, contextItems: List<ContextItem>) {
-    CodyAgentService.withAgentRestartIfNeeded(project) { agent ->
-      val message =
-          WebviewMessage(
-              command = "submit",
-              text = humanMessage.actualMessage(),
-              submitType = "user",
-              addEnhancedContext = chatPanel.isEnhancedContextEnabled(),
-              contextFiles = contextItems)
-
-      try {
-        val request =
-            agent.server.chatSubmitMessage(
-                ChatSubmitMessageParams(connectionId.get().get(), message))
-
-        GraphQlLogger.logCodyEvent(project, "chat-question", "submitted")
-
-        ApplicationManager.getApplication().invokeLater {
-          createCancellationToken(
-              onCancel = { request.cancel(true) },
-              onFinish = { GraphQlLogger.logCodyEvent(project, "chat-question", "executed") })
-        }
-      } catch (e: Exception) {
-        createCancellationToken(onCancel = {}, onFinish = {})
-        getCancellationToken().abort()
-        throw e
-      }
-    }
+    CodyAgentService.withAgentRestartIfNeeded(
+        project,
+        callback = { agent ->
+          val message =
+              WebviewMessage(
+                  command = "submit",
+                  text = humanMessage.actualMessage(),
+                  submitType = "user",
+                  addEnhancedContext = chatPanel.isEnhancedContextEnabled(),
```

Comment:
It's not for this PR but I just realized - there is something strange in our design. Our model logic (`submitMessageToAgent`) seems to depend on the view (the chat panel).